vllm/v1/attention/backends/utils.py

```diff
@@ -405,6 +407,49 @@ def make_local_attention_virtual_batches(
     )
 
 
+def subclass_attention_metadata_builder(
+    name_prefix: str,
+    builder_cls: type[AttentionMetadataBuilder[M]],
+    build_preprocess_fn: Callable[[CommonAttentionMetadata],
+                                  CommonAttentionMetadata],
+) -> type[AttentionMetadataBuilder[M]]:
+    """
+    Return a new subclass of `builder_cls` whose .build(...) method
+    first calls make_local_attention_virtual_batches(...) on the metadata.
+    """
+    name: str = name_prefix + builder_cls.__name__  # type: ignore
+
+    def build(self,
+              common_prefix_len: int,
+              common_attn_metadata: CommonAttentionMetadata,
+              fast_build: bool = False):
+        return builder_cls.build(self, common_prefix_len,
+                                 build_preprocess_fn(common_attn_metadata),
```

Comment:
should we also add a build_postprocess_fn? I think https://github.com/vllm-project/vllm/pull/21590 needs it for adding some special attribute to YOCO layers. Also CC @sarckk 
And @sarckk remeber to update https://github.com/vllm-project/vllm/pull/21590 to a cleaner way after this pr is landed.

Reply:
Sounds good!